test/Microsoft.AspNetCore.Server.KestrelTests/HttpParserTests.cs

```diff
@@ -111,7 +110,28 @@ public void ParseRequestLineThrowsOnInvalidRequestLine(string requestLine)
             var exception = Assert.Throws<BadHttpRequestException>(() =>
                 parser.ParseRequestLine(Mock.Of<IHttpRequestLineHandler>(), buffer, out var consumed, out var examined));
 
-            Assert.Equal($"Invalid request line: '{requestLine.Replace("\r", "\\x0D").Replace("\n", "\\x0A")}'", exception.Message);
+            Assert.Equal($"Invalid request line: '{requestLine.Replace("\r", "\\x0D").Replace("\n", "\\x0A").Replace("\0", "\\x00")}'", exception.Message);
+            Assert.Equal(StatusCodes.Status400BadRequest, (exception as BadHttpRequestException).StatusCode);
+        }
+
+        [Theory]
+        [MemberData(nameof(MethodWithNonTokenCharData))]
+        public void ParseRequestLineThrowsOnNonTokenCharsInCustomMethod(string method)
+        {
+            var requestLine = $"{method} / HTTP/1.1\r\n";
+
+            var mockTrace = new Mock<IKestrelTrace>();
+            mockTrace
+                .Setup(trace => trace.IsEnabled(LogLevel.Information))
+                .Returns(true);
+
+            var parser = CreateParser(mockTrace.Object);
+            var buffer = ReadableBuffer.Create(Encoding.ASCII.GetBytes(requestLine));
+
+            var exception = Assert.Throws<BadHttpRequestException>(() =>
+                parser.ParseRequestLine(Mock.Of<IHttpRequestLineHandler>(), buffer, out var consumed, out var examined));
+
+            Assert.Equal($"Invalid request line: '{method.Replace("\0", "\\x00")} / HTTP/1.1\\x0D\\x0A'", exception.Message);
```

Comment:
Can we add the character escaping to a common method? We've repeated it several times through the code.